internal/cli/image_clear_versioned.go

```diff
@@ -0,0 +1,179 @@
+// limitations under the License.
+
+package cli
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/torcx/internal/torcx"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var (
+	cmdImageClearVersioned = &cobra.Command{
+		Use:   "clear-versioned [INAME]",
+		Short: "clear images from versioned stores",
+		Long: `Clear images from versioned stores.
+If "INAME" is specified, only clear that image name from stores.`,
+		RunE: runImageClearVersioned,
+	}
+	flagImageClearVersionedKeepVersions []string
+
+	errNoStoreVersions = errors.New("no store versions to keep")
+	errNilCommonConfig = errors.New("nil CommongConfig")
+	errEmptyImageName  = errors.New("empty imageName")
+)
+
+func init() {
+	cmdImage.AddCommand(cmdImageClearVersioned)
+	cmdImageClearVersioned.Flags().StringArrayVarP(&flagImageClearVersionedKeepVersions, "keep-version", "k", []string{}, "store version to keep")
+}
+
+func runImageClearVersioned(cmd *cobra.Command, args []string) error {
+	keepVersions := flagImageClearVersionedKeepVersions
+	if len(keepVersions) == 0 {
+		return errNoStoreVersions
+	}
+	imageName := ""
+	if len(args) > 1 {
+		return errors.New("too many arguments")
+	}
+	if len(args) == 1 {
+		imageName = args[0]
+	}
+
+	commonCfg, err := fillCommonRuntime("")
+	if err != nil {
+		return errors.Wrap(err, "common configuration failed")
+	}
+
+	if imageName != "" {
+		if err := clearVersionedImage(commonCfg, imageName, keepVersions); err != nil {
+			return errors.Wrapf(err, "failed to clear versioned image %s", imageName)
+		}
+	} else {
+		if err := clearVersionedStores(commonCfg, keepVersions); err != nil {
+			return errors.Wrapf(err, "failed to clear versioned stores")
+		}
+	}
+
+	return nil
+}
+
+// clearVersionedImage deletes all references for the image `imageName` from all
+// versioned stores, except those listed in `keepVersions`.
+func clearVersionedImage(cfg *torcx.CommonConfig, imageName string, keepVersions []string) error {
+	if cfg == nil {
+		return errNilCommonConfig
+	}
+	if imageName == "" {
+		return errEmptyImageName
+	}
+
+	globPattern := filepath.Join(cfg.UserStorePath("*"), imageName+":*.torcx.*")
+	images, err := filepath.Glob(globPattern)
+	if err != nil {
+		return errors.Wrapf(err, "failed to glob on %s", globPattern)
+	}
+
+	removals := []string{}
+	for _, im := range images {
+		imPath := filepath.Clean(im)
+		fi, err := os.Stat(imPath)
+		if err != nil {
+			continue
+		}
+		if !fi.Mode().IsRegular() {
+			continue
+		}
+
+		keep := false
+		storePath := filepath.Dir(imPath)
+		for _, ver := range keepVersions {
+			verPath := filepath.Clean(cfg.UserStorePath(ver))
+			if keep || storePath == verPath {
+				keep = true
+				continue
+			}
+		}
+		if keep {
+			continue
+		}
+		removals = append(removals, imPath)
+	}
+
+	for _, imPath := range removals {
+		logrus.WithFields(logrus.Fields{
+			"path": imPath,
+		}).Warn("removing image from versioned store")
```

Comment:
Should that (and the other one in the next function) be a warning or just `Info`? We're doing what we were requested to do, right?